### PR TITLE
Fix `url_path_for` typehints

### LINKED ISSUE #1319: Inconsistent Typing Between request.url_for and application.url_path_for
[`request.url_for` takes `Any` for the `path_params`](https://github.com/encode/starlette/blob/217019b36303ea16a5869ad076d835e4fdc1cef3/starlette/requests.py#L168) whereas [`application.url_path_for` expects `path_params` to be `str`](https://github.com/encode/starlette/blob/537ab6afd110b79dca95f3c8ecc6980710b1de1c/starlette/applications.py#L107). Likewise the route / routers expect `path_params` to be `str`. See [here](https://github.com/encode/starlette/blob/537ab6afd110b79dca95f3c8ecc6980710b1de1c/starlette/routing.py#L159) and [here](https://github.com/encode/starlette/blob/537ab6afd110b79dca95f3c8ecc6980710b1de1c/starlette/routing.py#L584).

The type mismatch is not caught by the type checker as the `router` here is untyped:
https://github.com/encode/starlette/blob/217019b36303ea16a5869ad076d835e4fdc1cef3/starlette/requests.py#L169-L170

I think `Any` is probably more close to being the correct type as the `path_params` can be `UUID`, `int`, etc.

[The docs](https://www.starlette.io/routing/#reverse-url-lookups) would make it seems that the API for these two methods should have a consistent signature.

### PR DESCRIPTION
**Pull request by aminalaee**
Changes type hints for `url_path_for `. There's a mismatch between `url_for` and `url_path_for` as one accepts `Any` and the other one is `str`. They should both be `Any`.

Right now `url_path_for ` expects path_parameters to be string, which is not the case, and `url_for` could be `int`, `UUID`, `float`, etc.

Fixes #1319.

### CODE REVIEW

**Review by florimondmanca — APPROVED** [review]
From reviewing this and the associated issue, this seems legit to me!

File: starlette/requests.py [hunk 1]
```diff
@@ -15,6 +15,10 @@
     parse_options_header = None
 
 
+if typing.TYPE_CHECKING:  # pragma: nocover
```
Comment (PrettyWood) [inline starlette/requests.py]: IMO better to add directly in setup.cfg
```
[coverage:report]
exclude_lines =
    if typing.TYPE_CHECKING:
    @typing.overload
    ...
```
WDYT?

Comment (aminalaee) [reply starlette/requests.py]: Yeah makes sense. We'll probably need it in other places.

**Comment by aminalaee** [comment]
@PrettyWood @florimondmanca I think the `typing.Any` isn't worth much for the `path_params`.

Checking `starlette/convertors.py`:

```python
CONVERTOR_TYPES = {
    "str": StringConvertor(),
    "path": PathConvertor(),
    "int": IntegerConvertor(),
    "float": FloatConvertor(),
    "uuid": UUIDConvertor(),
}
``` 

we could have a `Union` of these to be more explicit. Any ideas?

File: setup.cfg [hunk 2]
```diff
@@ -32,3 +32,9 @@ filterwarnings=
 
 [coverage:run]
 source_pkgs = starlette, tests
+
+[coverage:report]
+exclude_lines =
+    pragma: no cover
+    pragma: nocover
```
Comment (Kludex) [inline setup.cfg:39]: Those 2 need to be included? :thinking:

Comment (aminalaee) [reply setup.cfg:39]: yes, overriding `exclude_lines` takes full control over what's excluded https://coverage.readthedocs.io/en/coverage-4.3.3/excluding.html.

**Pull request merged**